libblockchain/BlockChainImp.cpp

```diff
@@ -121,25 +121,78 @@ shared_ptr<TableFactory> BlockChainImp::getMemoryTableFactory(int64_t num)
     return memoryTableFactory;
 }
 
+std::shared_ptr<BlockHeaderInfo> BlockChainImp::getBlockHeaderInfo(int64_t _blockNumber)
+{
+    if (_blockNumber > number())
+    {
+        return nullptr;
+    }
+    // get block hash
+    auto blockHash = numberHash(_blockNumber);
+    if (blockHash != h256(""))
+    {
+        return getBlockHeaderInfoByHash(blockHash);
+    }
+    return nullptr;
+}
+
+std::shared_ptr<BlockHeaderInfo> BlockChainImp::getBlockHeaderFromBlock(dev::eth::Block::Ptr _block)
+{
+    if (!_block)
+    {
+        return nullptr;
+    }
+    // TODO: remove the copy overhead
+    return std::make_shared<BlockHeaderInfo>(
+        std::make_pair(std::make_shared<BlockHeader>(_block->blockHeader()), _block->sigList()));
+}
+
+std::shared_ptr<BlockHeaderInfo> BlockChainImp::getBlockHeaderInfoByHash(
+    dev::h256 const& _blockHash)
+{
+    auto cachedBlockInfo = m_blockCache.get(_blockHash);
+    // hit the cache, get block header from the cache directly
+    if (cachedBlockInfo.first)
+    {
+        return getBlockHeaderFromBlock(cachedBlockInfo.first);
+    }
+    // miss the cache, read from the SYS_HASH_2_BLOCKHEAER firstly
+    auto table = getMemoryTableFactory()->openTable(SYS_HASH_2_BLOCKHEADER);
+    if (!table)
+    {
+        // query block to obtain the block header and signature list
+        return getBlockHeaderFromBlock(getBlock(_blockHash));
+    }
+    auto entries = table->select(_blockHash.hex(), table->newCondition());
+    if (entries->size() <= 0)
+    {
+        return getBlockHeaderFromBlock(getBlock(_blockHash));
+    }
+    auto entry = entries->get(0);
+    // decode block header
+    auto blockHeaderBytes = entry->getFieldConst(SYS_VALUE);
+    auto blockHeader = std::make_shared<BlockHeader>(blockHeaderBytes, BlockDataType::HeaderData);
+
+    // decode signature list
+    auto sigListBytes = entry->getFieldConst(SYS_SIG_LIST);
+    auto sigList = std::make_shared<dev::eth::Block::SigListType>();
+    RLP rlp(sigListBytes);
+    *sigList = rlp.toVector<std::pair<u256, std::vector<unsigned char>>>();
+    return std::make_shared<BlockHeaderInfo>(std::make_pair(blockHeader, sigList));
+}
+
 std::shared_ptr<Block> BlockChainImp::getBlock(int64_t _blockNumber)
 {
     /// the future block
     if (_blockNumber > number())
     {
         return nullptr;
     }
-    Table::Ptr tb = getMemoryTableFactory(_blockNumber)->openTable(SYS_NUMBER_2_HASH);
-    if (tb)
+    auto blockHash = numberHash(_blockNumber);
```

Comment:
```suggestion
    auto blockHash = getBlockNumberOfHash(_blockNumber);
```
maybe more clear?

Reply:
`numberHash`  is a function that reuses the previous one, if you want to change it, you need to change it uniformly

Reply:
Maybe this name can be changed in the future PR.